consensus/event_loop.go

```diff
@@ -0,0 +1,854 @@
+
+	// Check if the event loop is no longer running. If it's not running
+	// we can simply return here.
+	if fc.status != eventLoopStatusRunning {
+		return
+	}
+
+	// Cancel the crank timer and timeout scheduled tasks, if any.
+	fc.crankTimerTask.Cancel()
+	fc.nextTimeoutTask.Cancel()
+
+	// Update the event loop's status so it is no longer marked as running.
+	fc.status = eventLoopStatusInitialized
+}
+
+func (fc *FastHotStuffEventLoop) IsInitialized() bool {
+	fc.lock.RLock()
+	defer fc.lock.RUnlock()
+
+	return fc.status != eventLoopStatusNotInitialized
+}
+
+func (fc *FastHotStuffEventLoop) IsRunning() bool {
+	fc.lock.RLock()
+	defer fc.lock.RUnlock()
+
+	return fc.status == eventLoopStatusRunning
+}
+
+// resetScheduledTasks recomputes the nextBlockConstructionTimeStamp and nextTimeoutTimeStamp
+// values, and reschedules the crank timer and timeout tasks.
+func (fc *FastHotStuffEventLoop) resetScheduledTasks() {
+	// Compute the next timeout ETA. We use exponential back-off for timeouts when there are
+	// multiple consecutive timeouts. We use the difference between the current view and the
+	// chain tip's view to determine this. The current view can only drift from the chain tip's
+	// view as a result of timeouts. This guarantees that the number of consecutive timeouts is
+	// always: max(currentView - tip.block.GetView() - 1, 0).
+
+	timeoutDuration := fc.timeoutBaseDuration
+
+	// Check if we have timed out at for the last n views. If so, we apply exponential
+	// back-off to the timeout base duration.
+	if fc.tip.block.GetView() < fc.currentView-1 {
+		// Note, there is no risk of underflow here because the following is guaranteed:
+		// currentView > tip.block.GetView() + 1.
+		numTimeouts := fc.currentView - fc.tip.block.GetView() - 1
+
+		// Compute the exponential back-off: nextTimeoutDuration * 2^numTimeouts
+		timeoutDuration = fc.timeoutBaseDuration << numTimeouts
+	}
+
+	// Schedule the next crank timer task. This will run with currentView param.
+	fc.crankTimerTask.Schedule(fc.crankTimerInterval, fc.currentView, fc.onCrankTimerTaskExecuted)
+
+	// Schedule the next timeout task. This will run with currentView param.
+	fc.nextTimeoutTask.Schedule(timeoutDuration, fc.currentView, fc.onTimeoutScheduledTaskExecuted)
+}
+
+// When this function is triggered, it means that we have reached the crank timer
+// time ETA for blockConstructionView. If we have a QC or timeout QC for the view, then we
+// signal the server.
+func (fc *FastHotStuffEventLoop) onCrankTimerTaskExecuted(blockConstructionView uint64) {
+	fc.lock.Lock()
+	defer fc.lock.Unlock()
+
+	// Check if the event loop is running. If it's not running, then there's nothing
+	// to do here.
+	if fc.status != eventLoopStatusRunning {
+		return
+	}
+
+	// Check for race conditions where the view advanced at the exact moment this task began
+	// or we have already signaled for this view. If so, then there's nothing to do here.
+	if fc.currentView != blockConstructionView {
```

Comment:
Smart. I thought I'd catch a bug related to this race condition but looks like you're on top of it.